Add Layout Conditions

MatyasKriz opened the pull request:
EBNF as of making this parser:
**CONDITION** := '[' EXPRESSION ']';
**EXPRESSION** := TERM [ or TERM ];
**TERM** := COMPARISON [ and COMPARISON ];
**COMPARISON** := FACTOR [ ( == | != | ':gt' | ':gte' | ':lt' | ':lte' ) FACTOR ];
**FACTOR** := [ '!' ] ( '(' EXPRESSION ')' | IDENTIFIER | FLOAT_NUMBER );
**FLOAT_NUMBER** := { same as Swift's `Float` }
**IDENTIFIER** := { all cases in the ConditionStatement enum };

Also in this PR:
- some changes I don't even know about,
- added `:gte` and `:lte` aliases to constraints to replace misleading `:gt` and `:lt`,
- some generator refactoring.

TadeasKriz committed:
Add LiveUI target and test target to ReactantUI.xcodeproj.

rkrenecky committed:
Add incomplete condition parsing with tests.

rkrenecky committed:
Simplify conditions logic and tests.

rkrenecky committed:
Add working implementation for condition parsing.

rkrenecky committed:
Little code refactoring.

rkrenecky committed:
Add condition parsing to separate file and fix bugs.

Commit:
Initial condition generation + added tests.

With the help of one and only @TadeasKriz.

Commit:
Add a guard to tests.

Commit:
Fix incorrect test.

Commit:
Implement condition parsing using recursive descent.

Commit:
Add “&&” to the example as `&amp;&amp;`.

Commit:
Call condition generating with `try!` until there’s a better way to show the error.

Commit:
Generate XML conditions. `&& -> and`, `|| -> or`.

Commit:
Have tests use new operators.

Commit:
Use new operators in example root view.

Commit:
Make `and` and `or` into reserved words. Moved above `identifier` in `Lexer` priorities.

Commit:
Fix EBNF.

Commit:
Add topview width and height comparison.

Commit:
Fix EBNF.

Commit:
Add new comparison operators to EBNF.

matoushybl commented on Example/Source/Component/SelectionCell/SelectionCell.swift:
```diff
@@ -30,7 +30,7 @@ enum ExampleType {
 }
 
-final class SelectionCell: ControlBase<ExampleType, ExampleType> {
+final class SelectionCell: ButtonBase<ExampleType, ExampleType> {
```
Why? Why is Reactant downgraded back to 1.1.0?

TadeasKriz replied:
Was the `ControlBase` released though?

MatyasKriz replied:
I have no memory of changing this.

TadeasKriz commented on Example/Source/Component/Playground/ExampleRootView.ui.xml:
```diff
@@ -22,7 +22,9 @@
         backgroundColor="blue"
         layout:id="test"
-        layout:leading="super"
+        layout:leading="[!(horizontal == compact) and (vertical == regular) and pad or width :lt 100] super;
```
`... or width.lessThan?(100)`

MatyasKriz replied:
If we're making the parser into function-parsing grade, we should decide on that now otherwise it will get uglier and uglier.

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -105,7 +129,65 @@ public class UIGenerator: Generator {
+                            }
+
+                            let name: String
+                            if child.field != nil {
```
What do you think about a method that would collect names of the children? Very similar code is used in `generateChildProperties `.

MatyasKriz replied:
I think the code might have been blatantly copied and slightly modified, that's why they're so similar.

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -105,7 +129,65 @@ public class UIGenerator: Generator {
+                        for guardName in guardNames {
+                            l("  let \(guardName) = self.\(guardName)\(guardName == guardNames.last ? "" : ",")")
+                        }
+                        l("else { /* FIXME Should we fatalError here? */ return }")
```
Can this happen in any other case than that the generator is broken?

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -105,7 +129,65 @@ public class UIGenerator: Generator {
+                            l("#if (arch(i386) || arch(x86_64)) && (os(iOS) || os(tvOS))")
+                        }
+                        // This will register `self` to remove `deinit` from ViewBase
+                        //l("ReactantLiveUIManager.shared.update(target)")
```
What about this dead code?

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -105,7 +129,65 @@ public class UIGenerator: Generator {
+                        l("#else")
+                    }
+
+                    // TODO: Add conditional properties?
```
What did these dead lines do?

MatyasKriz replied:
I'll be honest, it's quite a lot of time since we changed this file with @TadeasKriz and it's mainly his work in this file.

matoushybl replied:
No problem, as long as everything still works. I understand that this part of the project is a mess.

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -271,7 +363,14 @@ public class UIGenerator: Generator {
-        return constraintLine
+
+        if let condition = constraint.condition {
+            l("if \(try! constraint.generateSwiftCondition(condition: condition, viewName: name))") {
```
Generator should support exceptions now so the `try!` should not be needed.

MatyasKriz replied:
Wait, I thought I removed all of these pesky triers.

matoushybl commented on Sources/Tokenizer/Layout/Constraints/Constraint.swift:
```diff
@@ -101,6 +108,209 @@ public struct Constraint {
+    }
+
+    var xmlRepresentation: String {
+        switch self {
```
Why not return `swiftRepresentation` without switching?

MatyasKriz replied:
It's premature change for future maintainer when we wants to add new unary operator. I guess we can have a dumb repeater for now.

matoushybl reviewed (approved):
Nice work!

Commit:
Add `:gte` and `:lte` to both conditions and constraints.

Commit:
Rename rootView acquiring.

Commit:
Refactor `generatedNames` functions.

matoushybl commented on Sources/Generator/UIGenerator.swift:
```diff
@@ -105,7 +110,43 @@ public class UIGenerator: Generator {
+                l("func updateReactantUI()") {
+                    tempCounter = 1
+                    let guardNames = ["target"] + root.children.generatedNames(tempCounter: &tempCounter)
+                    if !guardNames.isEmpty {
```
Does this if make sense since there will always be at least `target`?

MatyasKriz replied:
Nice catch.

Commit:
Remove redundant `if-statement`.

Commit:
Prepare `attributedTextStyle`.

Commit:
Initial `attributedText` implementation.

Commit:
Make static variables into constants.

Commit:
Make `ControlState` conform to `EnumPropertyType`.

Commit:
Stylistic changes.

Commit:
Initial `Label.text` and `Button.title` parsing. Add `Array` to conform to `SupportedPropertyType`.

TadeasKriz committed:
WIP How element properties could work.

TadeasKriz committed:
WIP Add contexts.

Commit:
WIP context integration

TODO: Make everything conform to `AttributeSupportedPropertyType` instead of `SupportedPropertyType`.

Commit:
Add conformance to `AttributeSupportedPropertyType` to all implemented types.

Commit:
Add initial `AttributeText` setting.

Commit:
WIP `AttributedText` inheritance.

Commit:
Fix bug in attributed text parsing.

Commit:
Use `map` instead of `flatMap`.

Commit:
Add working inheritance attributed text styling.

Commit:
Finish static attributed text styling.

TODO: LiveUI can’t access the attributed text structures through names only
FIXME: `and` -> `&&` && `or` -> `||`

Commit:
Fix regression introduced in condition parsing.

`and` -> `&&`, `or` -> `||`
This commit also introduces case insensitive `and` and `or` inside conditions.

Commit:
Some formatting changes.

Commit:
Finish removing `and` and `or` tokens. Fix font defining in examples.

Commit:
Remove unused default values.

Commit:
Implement constraint condition validation.

Commit:
WIP Working conditions, XSD generating is crashy.

TadeasKriz committed:
Fix XSD crashing.

Commit:
Remove dev team.

Commit:
Remove useless computed variable.

Commit:
Replace `ReactantRuntime` with `canImport(UIKit)`.

Commit:
Implement component-wide attributed text styling.

Commit:
Implement stylegroup attributed text styling.

Commit:
Refactor resolved attributes from StyleGroup or component.

Commit:
Allow dot notation in style referencing.

Commit:
Initial attributed text extending.

Commit:
Fix wrong attribute order.

Commit:
Optimize attributed text appending.

Commit:
Trim trailing whitespace if the first element is also the last one.
